File: src/tools/semgrep/eval-repos/synthetic/java/SqlInjection.java
/**
 * Test file for SQL injection vulnerability detection.
 * Contains multiple SQL injection patterns.
 */

package smells.security;

import java.sql.*;

public class SqlInjection {

    // SQL INJECTION: String concatenation in SQL query
    public ResultSet getUserByName(Connection conn, String username) throws SQLException {
        String query = "SELECT * FROM users WHERE username = '" + username + "'";
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(query);
    }

    // SQL INJECTION: String format in SQL query
    public void deleteUser(Connection conn, String userId) throws SQLException {
        String query = String.format("DELETE FROM users WHERE id = %s", userId);
        Statement stmt = conn.createStatement();
        stmt.executeUpdate(query);
    }

    // SQL INJECTION: Multiple injectable parameters
    public ResultSet searchProducts(Connection conn, String name, String category) throws SQLException {
        String query = "SELECT * FROM products WHERE name LIKE '%" + name + "%' AND category = '" + category + "'";
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(query);
    }

    // SQL INJECTION: StringBuilder with user input
    public ResultSet getOrdersByDate(Connection conn, String startDate, String endDate) throws SQLException {
        StringBuilder sb = new StringBuilder();
        sb.append("SELECT * FROM orders WHERE created_at BETWEEN '");
        sb.append(startDate);
        sb.append("' AND '");
        sb.append(endDate);
        sb.append("'");
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(sb.toString());
    }

    // SQL INJECTION: Using concat
    public void updateEmail(Connection conn, String userId, String email) throws SQLException {
        String query = "UPDATE users SET email = '".concat(email).concat("' WHERE id = ").concat(userId);
        Statement stmt = conn.createStatement();
        stmt.executeUpdate(query);
    }

    // SAFE: Parameterized query
    public ResultSet safeGetUser(Connection conn, int userId) throws SQLException {
        PreparedStatement pstmt = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
        pstmt.setInt(1, userId);
        return pstmt.executeQuery();
    }

    // SAFE: Another parameterized query
    public void safeUpdateEmail(Connection conn, int userId, String email) throws SQLException {
        PreparedStatement pstmt = conn.prepareStatement("UPDATE users SET email = ? WHERE id = ?");
        pstmt.setString(1, email);
        pstmt.setInt(2, userId);
        pstmt.executeUpdate();
    }
}

File: src/tools/semgrep/eval-repos/synthetic/csharp/SqlInjection.cs
/**
 * Test file for SQL Injection vulnerability detection.
 * Contains multiple SQL injection patterns for C#.
 */

using System;
using System.Data;
using System.Data.SqlClient;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

namespace SmellTests.Security
{
    public class SqlInjectionExamples
    {
        private readonly string _connectionString = "Server=localhost;Database=test;";

        // SQL_INJECTION: String concatenation in SQL query
        public DataTable GetUserByName(string userName)
        {
            using var connection = new SqlConnection(_connectionString);
            // BAD: Direct string concatenation
            string query = "SELECT * FROM Users WHERE Name = '" + userName + "'";
            using var command = new SqlCommand(query, connection);
            var adapter = new SqlDataAdapter(command);
            var table = new DataTable();
            adapter.Fill(table);
            return table;
        }

        // SQL_INJECTION: String interpolation in SQL query
        public async Task<int> DeleteUser(int userId, string reason)
        {
            using var connection = new SqlConnection(_connectionString);
            // BAD: String interpolation
            string query = $"DELETE FROM Users WHERE Id = {userId} AND Reason LIKE '%{reason}%'";
            using var command = new SqlCommand(query, connection);
            await connection.OpenAsync();
            return await command.ExecuteNonQueryAsync();
        }

        // SQL_INJECTION: String.Format in SQL query
        public DataTable SearchProducts(string searchTerm, string category)
        {
            using var connection = new SqlConnection(_connectionString);
            // BAD: String.Format
            string query = string.Format(
                "SELECT * FROM Products WHERE Name LIKE '%{0}%' AND Category = '{1}'",
                searchTerm,
                category);
            using var command = new SqlCommand(query, connection);
            var adapter = new SqlDataAdapter(command);
            var table = new DataTable();
            adapter.Fill(table);
            return table;
        }

        // SQL_INJECTION: ExecuteSqlRaw with string concatenation
        public void UpdateOrderStatus(DbContext context, int orderId, string status)
        {
            // BAD: Raw SQL with concatenation in EF Core
            context.Database.ExecuteSqlRaw(
                "UPDATE Orders SET Status = '" + status + "' WHERE Id = " + orderId);
        }

        // SQL_INJECTION: Dynamic ORDER BY clause
        public DataTable GetSortedUsers(string sortColumn)
        {
            using var connection = new SqlConnection(_connectionString);
            // BAD: Dynamic column name (still injectable)
            string query = "SELECT * FROM Users ORDER BY " + sortColumn;
            using var command = new SqlCommand(query, connection);
            var adapter = new SqlDataAdapter(command);
            var table = new DataTable();
            adapter.Fill(table);
            return table;
        }

        // SQL_INJECTION: IN clause with user input
        public DataTable GetUsersByIds(string[] userIds)
        {
            using var connection = new SqlConnection(_connectionString);
            // BAD: Building IN clause from user input
            string idList = string.Join(",", userIds);
            string query = "SELECT * FROM Users WHERE Id IN (" + idList + ")";
            using var command = new SqlCommand(query, connection);
            var adapter = new SqlDataAdapter(command);
            var table = new DataTable();
            adapter.Fill(table);
            return table;
        }

        // CORRECT: Parameterized query
        public DataTable GetUserByNameSafe(string userName)
        {
            using var connection = new SqlConnection(_connectionString);
            // GOOD: Using parameters
            string query = "SELECT * FROM Users WHERE Name = @name";
            using var command = new SqlCommand(query, connection);
            command.Parameters.AddWithValue("@name", userName);
            var adapter = new SqlDataAdapter(command);
            var table = new DataTable();
            adapter.Fill(table);
            return table;
        }

        // CORRECT: Using stored procedure
        public DataTable GetUserByIdSafe(int userId)
        {
            using var connection = new SqlConnection(_connectionString);
            using var command = new SqlCommand("sp_GetUserById", connection);
            command.CommandType = CommandType.StoredProcedure;
            command.Parameters.AddWithValue("@userId", userId);
            var adapter = new SqlDataAdapter(command);
            var table = new DataTable();
            adapter.Fill(table);
            return table;
        }
    }
}
